tests/testthat/test-rust_result.R

```diff
@@ -0,0 +1,107 @@
+test_that("is_result", {
+
+  are_ok_result = list(
+    list(ok=NULL,err=NULL), #is, if both null, that encodes ok NULL
+    list(ok=42,err=NULL),
+    list(ok=NA,err=NULL),
+    list(ok=list(ok=NA,err=NULL),err=NULL), #result inside result
+    Ok(42),
+    Ok(Ok(42)),
+    Ok(Err(42))
+  )
+
+  are_err_result = list(
+    list(ok=NULL,err="42"),
+    list(ok=NULL,err=NA),
+    list(ok=NULL,err=list()),
+    Err("something went wrong"),
+    Err(Err(1)),
+    Err(Ok(NULL))
+  )
+
+  are_invalid_result = list(
+    list(ok=42,err=42),#not if both ok err has value
+    list(err=NULL,ok=42), #not if flip order ok, err
+    list(ok=list(),err=list()),
+    list(list(ok=42,err=NULL)),
+    42,
+    NULL,
+    list(),
+    NA
+  )
+
+
+  are_results = c(
+    are_ok_result,
+    are_err_result
+  )
+
+  not_results = list(
+    are_invalid_result
+  )
+
+  #test is_result
+  for(i in are_results) expect_true( is_result(i), info = paste("testcase was ",str_string(i)))
```

Comment:
It may be worth considering using patrick.
https://github.com/google/patrick

Reply:
I will try it out next time :)

Reply:
I'm not sure this rust_result should be taken too far. It will probably push non rust R users away from reading the source code. What do you think?

Reply:
I do not think this test is hard to read.

Reply:
how about this ?
https://github.com/pola-rs/r-polars/blob/d8ad7d9a99b169be5581f4856f3bef1d4bcc93fd/R/expr__string.R#L57

it allows me to append context to err messages without a package. But maybe it is too strange to read for non rustaceans

Reply:
Oh, I don't know what the `pcase` function is doing......
Perhaps something like switch or case when?

Reply:
yes exactly it is a rip-off from data.table::fcase, p for polars

https://github.com/pola-rs/r-polars/blob/fd3dcb5f92e316611597c5fb080ce44d30dba1f1/R/utils.R#L114